packages/docusaurus-mdx-loader/src/remark/toc/search.ts

```diff
@@ -8,40 +8,83 @@
 import toString from 'mdast-util-to-string';
 import visit, {Visitor} from 'unist-util-visit';
 import {toValue} from '../utils';
-import type {TOCItem as TOC} from '@docusaurus/types';
+import type {TOCItem} from '@docusaurus/types';
 import type {Node} from 'unist';
 import type {Heading} from 'mdast';
 
-// Visit all headings. We `slug` all headings (to account for
-// duplicates), but only take h2 and h3 headings.
-export default function search(node: Node): TOC[] {
-  const headings: TOC[] = [];
-  let current = -1;
-  let currentDepth = 0;
+// Intermediate interface for TOC algorithm
+interface SearchItem {
+  node: TOCItem;
+  level: number;
+  parentIndex: number;
+}
+
+/**
+ *
+ * Generate a TOC AST from the raw Markdown contents
+ */
+export default function search(node: Node): TOCItem[] {
+  const headings: SearchItem[] = [];
 
   const visitor: Visitor<Heading> = (child, _index, parent) => {
     const value = toString(child);
 
-    if (parent !== node || !value || child.depth > 3 || child.depth < 2) {
+    // depth:1 headings are titles and not included in the TOC
+    if (parent !== node || !value || child.depth < 2) {
       return;
     }
 
-    const entry: TOC = {
-      value: toValue(child),
-      id: child.data!.id as string,
-      children: [],
-    };
+    headings.push({
+      node: {
+        value: toValue(child),
+        id: child.data!.id as string,
+        children: [],
+      },
+      level: child.depth,
+      parentIndex: -1,
+    });
+  };
 
-    if (!headings.length || currentDepth >= child.depth) {
-      headings.push(entry);
-      current += 1;
-      currentDepth = child.depth;
-    } else {
-      headings[current].children.push(entry);
+  visit(node, 'heading', visitor);
+
+  const getParentIndex = (prevParentIndex: number, currLevel: number) => {
+    let parentIndex = prevParentIndex;
+    // We start at the parent of the previous heading
+    // Recurse through its ancestors until the current heading would be a child
+    while (parentIndex >= 0 && currLevel < headings[parentIndex].level) {
+      parentIndex = headings[parentIndex].parentIndex;
     }
+    return parentIndex >= 0 ? headings[parentIndex].parentIndex : parentIndex;
   };
 
-  visit(node, 'heading', visitor);
+  // Assign the correct `parentIndex` for each heading.
+  headings.forEach((curr, currIndex) => {
+    if (currIndex > 0) {
+      const prevIndex = currIndex - 1;
+      const prev = headings[prevIndex];
+      if (curr.level > prev.level) {
+        curr.parentIndex = prevIndex;
+      } else if (curr.level < prev.level) {
+        curr.parentIndex = getParentIndex(prev.parentIndex, curr.level);
+      } else {
+        curr.parentIndex = prev.parentIndex;
+      }
+    }
+  });
```

Comment:
I see a better solution here: record "the nearest `h2`~`h6` node indices" as an array `closestNode[]`, and thus `curr.parentIndex = closestNode[curr.level - 1]`. And then you update by `closestNode[curr.level] = currIndex`

Reply:
Just realized there's one weird case where my algorithm doesn't work:

```
## h2
#### h4
### h3
##### h5
```

Maybe something that says "find the value in `closestNode` with a smaller level than `curr.level` and the largest index possible?